src/tests/testCbcPointer.cc

```diff
@@ -0,0 +1,260 @@
+public:
+    Blob() { ++Instances; }
+    ~Blob() { --Instances; }
+
+    static int Instances;
+};
+CBDATA_CLASS_INIT(Blob);
+
+int Blob::Instances = 0;
+
+void
+TestCbcPointer::testConstruction()
+{
+    Blob *nil = nullptr;
+    {
+        // default nil
+        CbcPointer<Blob> empty;
+        CPPUNIT_ASSERT_EQUAL(nil, empty.raw());
+        CPPUNIT_ASSERT_EQUAL(false, empty.set());
+        CPPUNIT_ASSERT_EQUAL(nil, empty.get());
+        // destruct of default-nil
+    }
+
+    {
+        // explicit nil
+        CbcPointer<Blob> empty(nullptr);
+        CPPUNIT_ASSERT_EQUAL(nil, empty.raw());
+        CPPUNIT_ASSERT_EQUAL(false, empty.set());
+        CPPUNIT_ASSERT_EQUAL(nil, empty.get());
+        // destruct of explicit-nil
+    }
+
+    {
+        auto expect = new Blob;
+        // delayed assignment from raw-pointer
+        CbcPointer<Blob> ptr(expect);
+        CPPUNIT_ASSERT_EQUAL(expect, ptr.raw());
+        CPPUNIT_ASSERT_EQUAL(true, ptr.set());
+        CPPUNIT_ASSERT_EQUAL(expect, ptr.get());
+
+        // invalidated by delete operator
+        delete ptr.get();
+        // memory should remain valid while ptr is set
+        CPPUNIT_ASSERT_EQUAL(static_cast<void*>(expect), expect->toCbdata());
+        CPPUNIT_ASSERT_EQUAL(expect, ptr.raw());
+        CPPUNIT_ASSERT_EQUAL(true, ptr.set());
+        CPPUNIT_ASSERT_EQUAL(nil, ptr.get());
+        // destruct of dirty Pointer
+    }
+
+    // copy and move construction are checked in testCounting()
+}
+
+void
+TestCbcPointer::testLogicOperators()
+{
+    // nil pointer semantics
+    {
+        CbcPointer<Blob> ptr;
+        CbcPointer<Blob> other;
+        CPPUNIT_ASSERT_EQUAL(static_cast<Blob *>(nullptr), ptr.valid());
+        CPPUNIT_ASSERT_EQUAL(true, !ptr);
+        CPPUNIT_ASSERT(other == ptr);
+    }
+
+    // valid pointer semantics
+    {
+        auto expect = new Blob;
+        CbcPointer<Blob> ptr(expect);
+        CbcPointer<Blob> same(expect);
+        CbcPointer<Blob> other;
+        CPPUNIT_ASSERT_EQUAL(expect, ptr.valid());
+        CPPUNIT_ASSERT_EQUAL(false, !ptr);
+        CPPUNIT_ASSERT(same == ptr);
+        // TODO: missing operator !=()
+        CPPUNIT_ASSERT(!(other == ptr));
+        // CbcPointer does not invalidate on destruct. prevent a leak
+        delete expect;
+    }
+
+    // invalid pointer semantics
+    {
+        auto expect = new Blob;
+        CbcPointer<Blob> ptr(expect);
+        CbcPointer<Blob> same(expect);
+        CbcPointer<Blob> other;
+        delete expect; // invalidates
+        CPPUNIT_ASSERT_EQUAL(static_cast<Blob *>(nullptr), ptr.valid());
+        CPPUNIT_ASSERT_EQUAL(true, !ptr);
+        CPPUNIT_ASSERT(same == ptr);
+        // TODO: missing operator !=()
+        CPPUNIT_ASSERT(!(other == ptr));
+    }
+}
+
+void
+TestCbcPointer::testCounting()
+{
+    // check no leaks from earlier tests
+    CPPUNIT_ASSERT_EQUAL(0, Blob::Instances);
+
+    // counting with nil pointer should never allocate
```

Comment:
Well, this is testing for allocations. The constructor is used as a reasonable proxy for allocate. I consider it such because the unit test stub_libmem allocator intentionally uses ```xalloc()/xfree()``` directly to avoid any pooling layer complications. Every CBDATA call to the constructor/destructor maps to exactly one call to xalloc/xfree respectively.

Reply:
> The constructor is used as a reasonable proxy for allocate

Please do not use indirect terminology where it is not necessary and where others find it confusing/misleading (especially when others are experts in the corresponding area). PR code that counts Blob construction and destruction calls (and feels the need to comment about it) should call them construction/destruction calls rather than "allocations", especially since Blob constructor and destructor do not allocate/deallocate.

> Every CBDATA call to the constructor/destructor maps to exactly one call to xalloc/xfree respectively.

This PR tests CbcPointer<Blob>. CbcPointer<Blob> never calls Blob constructor/destructor (directly or indirectly). 

CbcPointer does (indirectly) call xalloc/xfree or equivalent, but those calls have very little to do with Blob constructor/destructor in the current CBDATA implementation and may have _nothing_ to do with Blob constructor/destructor in other valid CBDATA implementations (the exact relationship is a CBDATA implementation-defined detail that this code should not know about).